File: sdks/python/taskito/dashboard/handlers/overrides.py
"""Task & queue override endpoints."""

from __future__ import annotations

from dataclasses import asdict
from typing import TYPE_CHECKING, Any

from taskito.dashboard.errors import _BadRequest, _NotFound
from taskito.dashboard.overrides_store import (
    QUEUE_OVERRIDE_FIELDS,
    TASK_OVERRIDE_FIELDS,
    OverridesStore,
)

if TYPE_CHECKING:
    from taskito.app import Queue


def handle_list_tasks(queue: Queue, _qs: dict) -> list[dict[str, Any]]:
    """Return every registered task with decorator defaults + active override."""
    return queue.registered_tasks()


def handle_list_queues(queue: Queue, _qs: dict) -> list[dict[str, Any]]:
    return queue.registered_queues()


def _coerce_override_body(body: Any, allowed: frozenset[str]) -> dict[str, Any]:
    if not isinstance(body, dict):
        raise _BadRequest("body must be a JSON object")
    unknown = set(body) - allowed
    if unknown:
        raise _BadRequest(
            f"unknown override fields: {sorted(unknown)}; allowed: {sorted(allowed)}"
        )
    return body


# ── Task override endpoints ───────────────────────────────────────────


def handle_get_task_override(queue: Queue, _qs: dict, task_name: str) -> dict[str, Any]:
    override = OverridesStore(queue).get_task(task_name)
    if override is None:
        raise _NotFound(f"no override set for task '{task_name}'")
    return asdict(override)


def handle_put_task_override(queue: Queue, body: dict, task_name: str) -> dict[str, Any]:
    fields = _coerce_override_body(body, TASK_OVERRIDE_FIELDS)
    try:
        override = OverridesStore(queue).set_task(task_name, fields)
    except ValueError as e:
        raise _BadRequest(str(e)) from None
    return asdict(override)


def handle_delete_task_override(queue: Queue, task_name: str) -> dict[str, bool]:
    removed = OverridesStore(queue).clear_task(task_name)
    return {"cleared": removed}


# ── Queue override endpoints ──────────────────────────────────────────


def handle_get_queue_override(queue: Queue, _qs: dict, queue_name: str) -> dict[str, Any]:
    override = OverridesStore(queue).get_queue(queue_name)
    if override is None:
        raise _NotFound(f"no override set for queue '{queue_name}'")
    return asdict(override)


def handle_put_queue_override(queue: Queue, body: dict, queue_name: str) -> dict[str, Any]:
    fields = _coerce_override_body(body, QUEUE_OVERRIDE_FIELDS)
    try:
        override = OverridesStore(queue).set_queue(queue_name, fields)
    except ValueError as e:
        raise _BadRequest(str(e)) from None
    # Reflect "paused" immediately by touching the paused_queues store
    # (this state DOES propagate to a running worker — independent of the
    # static override consumed at worker startup).
    if "paused" in fields:
        try:
            if fields["paused"]:
                queue.pause(queue_name)
            else:
                queue.resume(queue_name)
        except Exception:  # pragma: no cover - safety net only
            pass
    return asdict(override)


def handle_delete_queue_override(queue: Queue, queue_name: str) -> dict[str, bool]:
    removed = OverridesStore(queue).clear_queue(queue_name)
    return {"cleared": removed}

File: sdks/node/src/dashboard/handlers/overrides.ts
// Task & queue override endpoints.

import type { Queue } from "../../index";
import { BadRequestError, NotFoundError } from "../errors";

/** Every registered task with registration defaults + active override. */
export function listTasks(queue: Queue) {
  return queue.registeredTasks();
}

export function listQueues(queue: Queue) {
  return queue.registeredQueues();
}

function requireObject(body: unknown): Record<string, unknown> {
  if (!body || typeof body !== "object" || Array.isArray(body)) {
    throw new BadRequestError("body must be a JSON object");
  }
  return body as Record<string, unknown>;
}

// ── Task overrides ──────────────────────────────────────────────────────

export function getTaskOverride(queue: Queue, taskName: string) {
  const override = queue.getTaskOverride(taskName);
  if (!override) {
    throw new NotFoundError(`no override set for task '${taskName}'`);
  }
  return override;
}

// Store-level ValidationError propagates to the server, which maps it to 400.
export function putTaskOverride(queue: Queue, taskName: string, body: unknown) {
  return queue.setTaskOverride(taskName, requireObject(body));
}

export function deleteTaskOverride(queue: Queue, taskName: string) {
  return { cleared: queue.clearTaskOverride(taskName) };
}

// ── Queue overrides ─────────────────────────────────────────────────────

export function getQueueOverride(queue: Queue, queueName: string) {
  const override = queue.getQueueOverride(queueName);
  if (!override) {
    throw new NotFoundError(`no override set for queue '${queueName}'`);
  }
  return override;
}

export function putQueueOverride(queue: Queue, queueName: string, body: unknown) {
  const fields = requireObject(body);
  const override = queue.setQueueOverride(queueName, fields);
  // "paused" propagates to running workers immediately via the paused-queues
  // store, independent of the static override consumed at worker startup.
  if ("paused" in fields) {
    if (fields.paused) {
      queue.pauseQueue(queueName);
    } else {
      queue.resumeQueue(queueName);
    }
  }
  return override;
}

export function deleteQueueOverride(queue: Queue, queueName: string) {
  return { cleared: queue.clearQueueOverride(queueName) };
}
